docs/features/software-catalog/external-integrations.md

```diff
@@ -9,24 +9,282 @@ Backstage natively supports importing catalog data through the use of
+  to do their job; in that case you'll have to make that happen.
+- The timing of their work is entirely detached from the processing loops. One
+  provider may run every 30 seconds, another one on every incoming webhook call
+  of a certain type, etc.
+- They can perform detailed updates on the set of entities that they are
+  responsible for. They can make full updates of the entire set, or issue
+  individual additions and removals.
+- Their output is a set of unprocessed entities. Those are then subject to the
+  processing loops before becoming final, stitched entities.
+- When they remove an entity, the entire subtree of processor-generated entities
+  under that root is eagerly removed as well.
+
+### Creating an Entity Provider
 
-As a Backstage adopter, you would be able to customize or extend the catalog in
-several ways - by replacing the entire backend API, or by replacing entire
-implementation classes at certain points in the backend, or by leveraging the
-ingestion process to fetch data from your own authoritative source. Each method
-has benefits and drawbacks, but this article will focus on the last one of the
-above. It is the one that is the most straight forward and future proof, and
-leverages a lot of benefits that come with the builtin catalog.
+The recommended way of instantiating the catalog backend classes is to use the
+`CatalogBuilder`, as illustrated in the
+[example backend here](https://github.com/backstage/backstage/blob/master/packages/backend/src/plugins/catalog.ts).
+We will create a new
+[`EntityProvider`](https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/providers/types.ts)
+subclass that can be added to this catalog builder.
+
+Let's make a simple provider that can refresh a pool of entities based on a
+remote store. The provider part of the interface is actually tiny - you only
+have to supply a (unique) name, and accept a connection from the environment
+through which you can issue writes. The rest is up to the individual provider
+implementation.
+
+It is up to you where you put the code for this new processor class. For quick
+experimentation you could place it in your backend package, but we recommend
+putting all extensions like this in a backend plugin package of their own in the
+`plugins` folder of your Backstage repo.
+
+The class will have this basic structure:
+
+```ts
+import { UrlReader } from '@backstage/backend-common';
+import { Entity } from '@backstage/catalog-model';
+import {
+  EntityProvider,
+  EntityProviderConnection,
+} from '@backstage/plugin-catalog-backend';
+
+/**
+ * Provides entities from fictional frobs service.
+ */
+export class FrobsProvider implements EntityProvider {
+  private readonly env: string;
+  private readonly reader: UrlReader;
+  private connection?: EntityProviderConnection;
+
+  /** [1] **/
+  constructor(env: string, reader: UrlReader) {
+    this.env = env;
+    this.reader = reader;
+  }
+
+  /** [2] **/
```

Comment:
I'm somewhat leaning towards inlining the explanation here but 🤷

Reply:
Yeah definitely a valid idea! Was weighing pros and cons, it does keep things better collected but also implies getting a huge code block that breaks up reading even more, with monospaced font that's arguably worse for reading and has no support for rendering with emphasis, underlines, links etc.